src/accumulator/proof.rs

```diff
@@ -241,6 +243,286 @@ impl Proof {
+        let new_forest_rows = util::tree_rows(num_leaves + num_adds);
+        let old_forest_rows = util::tree_rows(num_leaves);
+        let tree_rows = util::tree_rows(num_leaves);
+        let mut new_proofs = vec![];
+        if new_forest_rows > old_forest_rows {
+            for (pos, hash) in positions.iter() {
+                let row = util::detect_row(*pos, tree_rows);
+
+                let old_start_pos = util::start_position_at_row(row, old_forest_rows);
+                let new_start_pos = util::start_position_at_row(row, new_forest_rows);
+
+                let offset = pos - old_start_pos;
+                let new_pos = offset + new_start_pos;
+                new_proofs.push((new_pos, *hash));
+            }
+            return new_proofs;
+        }
+
+        positions
+    }
+
+    /// update_proof_remove modifies the cached proof with the deletions that happen in the block proof.
+    /// It updates the necessary proof hashes and un-caches the targets that are being deleted.
+    fn update_proof_remove(
+        self,
+        block_targets: Vec<u64>,
+        cached_hashes: Vec<sha256::Hash>,
+        updated: Vec<(u64, sha256::Hash)>,
+        num_leaves: u64,
+    ) -> Result<(Proof, Vec<sha256::Hash>), String> {
+        let total_rows = util::tree_rows(num_leaves);
+
+        let targets_with_hash: Vec<(u64, bitcoin_hashes::sha256::Hash)> = self
+            .targets
+            .clone()
+            .into_iter()
+            .zip(cached_hashes.clone().into_iter())
+            .filter(|(pos, _)| !block_targets.contains(pos))
+            .collect();
+        let (targets, target_hashes) = targets_with_hash.to_owned().into_iter().unzip();
+
+        let proof_positions =
+            util::get_proof_positions(&self.targets, num_leaves, util::tree_rows(num_leaves));
+
+        let old_proof: Vec<_> = proof_positions
+            .iter()
+            .zip(self.hashes.clone().into_iter())
+            .collect();
+
+        let mut new_proof = vec![];
+        // Grab all the positions of the needed proof hashes.
+        let needed_pos = util::get_proof_positions(&targets, num_leaves, total_rows);
+
+        let updated: Vec<_> = updated
+            .into_iter()
+            .filter(|(pos, _)| proof_positions.contains(pos))
+            .collect();
+
+        let old_proof_iter = old_proof.into_iter();
+
+        // Loop through old_proofs and only add the needed proof hashes.
+        for (pos, hash) in old_proof_iter {
+            if let Some((_, updated_hash)) =
+                updated.iter().find(|(updated_pos, _)| pos == updated_pos)
+            {
+                if *updated_hash != sha256::Hash::default() {
+                    new_proof.push((*pos, *updated_hash));
+                }
+            } else {
+                if hash != sha256::Hash::default() {
+                    new_proof.push((*pos, hash));
+                }
+            }
+        }
+
+        let missing_positions = needed_pos
```

Comment:
`missing_positions` are the resulting vector of `u64`s from `get_proof_positions` minus the following:

1. Deleted targets in the current block. These would be `block_targets`.
2. All the current positions we're aware of. These would be included in `new_proof`. (I realized the go impl basically uses the result from proof_positions. This still would work since `new_proof` is a subset of `proof_positions` but we should be using `new_proof`).

We're not getting rid of the `block_targets` and that's why we're getting duplicates. Should add a filter for `block_targets` here.

Reply:
I think the problem was the filter closure. I should be: 
```rust
            .filter(|pos| !proof_positions.contains(pos));
```
But I missed the negation of `proof_positions.contains`, this brings elements that already are in the proof to `missing`, which is wrong.